### PR TITLE
Using only one form in linklist.html - fix #885

### LINKED ISSUE #885: The search-buttons in linklist.html are mutually exclusive
This is a bug report.

When a user clicks on either one of the magnifying glass icons at the top of the main view, in the search form, the other search input is ignored and discarded :

2017-05-30 16_15_19-liens utiles et a partager

For example, with `foo` in the text input and `bar` in the tags input :
- when clicking on the (left) text button, the tag query term is discarded
- when clicking on the (right) tag button, the text query term is discarded

I can send a PR to fix this.

Do you agree this is a bug ?

### PR DESCRIPTION
**Pull request by Lucas-C**
An instance displaying the effect of those changes can be tested here: https://chezsoi.org/shaarli

### CODE REVIEW

File: tpl/default/linklist.html [hunk 1]
```diff
@@ -19,30 +19,23 @@
...
-        autocomplete="off" data-multiple data-autofirst data-minChars="1"
-        data-list="{loop="$tags"}{$key}, {/loop}"
-        >
-        <button type="submit" class="search-button"><i class="fa fa-search"></i></button>
-      </form>
-    </div>
-  </div>
+  <form method="GET" class="searchform" name="searchform" class="pure-form">
+    <fieldset>
+      <input type="text" tabindex="1" name="searchterm" class="searchterm" laceholder="{'Search text'|t}"
```
Comment (nodiscc) [inline tpl/default/linklist.html]: typo: `laceholder`

File: index.php [hunk 2]
```diff
@@ -1425,7 +1425,7 @@ function renderPage($conf, $pluginManager, $LINKSDB, $history)
                 'url' => $url,
                 'description' => $description,
                 'tags' => $tags,
-                'private' => $private
+                'private' => $private,
```
Comment (virtualtam) [inline index.php]: unneeded change

File: tpl/default/linklist.html [hunk 3]
```diff
@@ -19,30 +19,23 @@
...
+             {if="!empty($search_term)"}
+             value="{$search_term}"
+             {/if}
+      >
+      <button type="submit" class="search-button"><i class="fa fa-search"></i></button>
```
Comment (virtualtam) [inline tpl/default/linklist.html]: having the search button between text input fields feels a bit weird, it would make more sense to place it after both input fields

File: tpl/default/linklist.html [hunk 4]
```diff
@@ -19,30 +19,23 @@
...
-               value="{$search_tags}"
-               {/if}
-        autocomplete="off" data-multiple data-autofirst data-minChars="1"
-        data-list="{loop="$tags"}{$key}, {/loop}"
-        >
-        <button type="submit" class="search-button"><i class="fa fa-search"></i></button>
-      </form>
-    </div>
-  </div>
+  <form method="GET" class="searchform" name="searchform" class="pure-form">
```
Comment (virtualtam) [inline tpl/default/linklist.html]: `class` defined twice

File: tpl/default/page.header.html [hunk 5]
```diff
@@ -97,30 +97,23 @@
...
-                 value="{$search_tags}"
-                 {/if}
-          autocomplete="off" data-multiple data-autofirst data-minChars="1"
-          data-list="{loop="$tags"}{$key}, {/loop}"
-          >
-          <button type="submit" class="search-button"><i class="fa fa-search"></i></button>
-        </form>
-      </div>
-    </div>
+    <form method="GET" class="searchform" name="searchform" class="pure-form">
```
Comment (virtualtam) [inline tpl/default/page.header.html]: `class` used twice

Comment (virtualtam) [inline tpl/default/page.header.html]: @ArthurHoaro what is the usage of having a hidden link search form included in all pages?

Comment (Lucas-C) [inline tpl/default/page.header.html]: I'm not sur I understand your question. Can you rephrase it ?
Because I did not ADD this form, it juste updated it.

Comment (virtualtam) [inline tpl/default/page.header.html]: yup, this is why I've tagged Arthur in the review comment ;-)

Comment (virtualtam) [inline tpl/default/page.header.html]: @Lucas-C this template - `page.header.html` is included in other pages' templates, and is consequently rendered for most Shaarli pages; as it generates a hidden form, and the search form is (currently) only displayed and usable from the `linklist` page, I'm wondering whether it is actually used, or should be removed from this template file

File: index.php (hunk 2 above)
Comment (Lucas-C) [inline index.php]: Fixed

File: tpl/default/linklist.html (hunk 4 above)
Comment (Lucas-C) [inline tpl/default/linklist.html]: Fixed

File: tpl/default/page.header.html (hunk 5 above)
Comment (Lucas-C) [inline tpl/default/page.header.html]: Fixed

File: tpl/default/linklist.html (hunk 3 above)
Comment (Lucas-C) [inline tpl/default/linklist.html]: Fixed

File: tpl/default/linklist.html (hunk 1 above)
Comment (Lucas-C) [inline tpl/default/linklist.html]: Well spotted ! Fixed

File: tpl/default/page.header.html (hunk 5 above)
Comment (Lucas-C) [inline tpl/default/page.header.html]: Ok ! Thanks for the explanation

**Comment by Lucas-C** [comment]
After some tests I've found an issue when an empty "searchtags" param is sent:
`!empty()` should be used instead of `isset()` here: https://github.com/shaarli/Shaarli/blob/master/application/LinkFilter.php#L70

I'm going to update the PR

**Comment by Lucas-C** [comment]
Could this last change be an issue for https://github.com/shaarli/Shaarli/issues/784 ?

**Comment by Lucas-C** [comment]
I made this change because when I filter the linklist with a tag, the remove it, an empty "searchtags" query param is sent.

File: tpl/default/page.header.html (hunk 5 above)
Comment (ArthurHoaro) [inline tpl/default/page.header.html]: @virtualtam In the default theme, the search is available everywhere from the search icon in the header.

**Comment by ArthurHoaro** [comment]
@Lucas-C You're right, your PR is in conflict with #784 / #841. Searching for an empty tag will result in searching in untagged links - which will happen for any text only search using a single form (as you did here).
A solution would be to remove the untagged feature using the search, and adding it back with a filter instead (an icon next to "private only" filter). What do you guys think?

Also, I don't understand why my commit from #882 is here.

**Comment by Lucas-C** [comment]
For the extra commit, that was an error, I removed it.

I think like your idea.
Just to make sure I understand it right: this extra icon would be a "non tagged links only" filter ?
Sounds great :)

Do you have any suggestion for the icon ?
And I think if this filter is active, the "tagfilter" input should be disabled.

**Comment by ArthurHoaro** [comment]
> Just to make sure I understand it right: this extra icon would be a "non tagged links only" filter ?

Yes.

> Do you have any suggestion for the icon ?

Not sure. Maybe the [Font Awesome tag](http://fontawesome.io/icon/tag/) enabled by default?

> And I think if this filter is active, the "tagfilter" input should be disabled.

Maybe. Or it could be disabled if a tag filter is requested.

Do you want to work on it in a separate PR?

**Comment by Lucas-C** [comment]
Yes ! :)
Can you merge this PR beforehand ? Else I'm going to get lost between "in progresse" PRs...

**Comment by ArthurHoaro** [comment]
Yup, but I'll review it first.

File: tpl/default/css/shaarli.css [hunk 6]
```diff
@@ -256,23 +256,17 @@ body, .pure-g [class*="pure-u"] {
...
-    #search .tagfilter,
-    #search-linklist .tagfilter {
-        margin-left: 25px;
-        text-align: left;
-    }
+#search,
+#search-linklist {
+    padding: 6px 0;
 }
+
```
Comment (ArthurHoaro) [inline tpl/default/css/shaarli.css]: I'm not sure what the part you removed was supposed to do, but there are 2 CSS issues:

  * with this padding, the header search is slightly overflowing the green bar

shaarli-search

  * with only one search icon the alignment is broken on mobile view

search-align-mobile

My suggestion for the latter would be to either:
  * remove the search icon (which I think is not very good for accessibility)
  * use an alternative block for mobile view using the grid system, something like (I didn't test it, just to give you an idea)

```html
<div class="pure-u-lg-0>
  <div class="pure-u-{large part of the screen}>
    {both fields}
  </div>
  <div class="pure-u-{small part of the screen}>
    {the icon}
  </div>
</div>
```

File: application/LinkDB.php [hunk 7]
```diff
@@ -430,7 +430,7 @@ public function filterSearch($filterRequest = array(), $casesensitive = false, $
         $type = LinkFilter::$FILTER_TAG | LinkFilter::$FILTER_TEXT;
         $request = [$searchtags, $searchterm];
 
-        $linkFilter = new LinkFilter($this);
+        $linkFilter = new LinkFilter($this->links);
```
Comment (ArthurHoaro) [inline application/LinkDB.php]: Is there any reason to change this? By passing the array instead of `$this`, `ArrayAccess` rules won't apply, i.e. mapping link IDs instead of array IDs, which might have side effects.

File: tpl/default/css/shaarli.css [hunk 8]
```diff
@@ -256,23 +256,17 @@ body, .pure-g [class*="pure-u"] {
     color: #d0d0d0;
 }
 
-#search-linklist {
-    padding: 5px 0;
+#search fieldset,
+#search-linklist fieldset {
+    border: none;
+    -webkit-padding-before: 0;
```
Comment (Lucas-C) [inline tpl/default/css/shaarli.css]: Fair point, I should have added a comment.
I replaced that by a simple `padding: 0`

File: application/LinkDB.php (hunk 7 above)
Comment (Lucas-C) [inline application/LinkDB.php]: Oh, ok, I did not realize that.
I'm going to drop this change.

File: tpl/default/css/shaarli.css (hunk 6 above)
Comment (Lucas-C) [inline tpl/default/css/shaarli.css]: - I don't see the padding issue under neither Firefox 53.0.2 nor Chrome 58.0.3029.110
What is your browser ?

- if I understand correctly, you would like the magnifying glass icon to be on a 3rd row in mobile view ? That's OK for me, I'm just not completly sure I got your point.

About accessibility : why wouldn't the search submit button be accessible ?
I'm genuinely curious here :)
It seems to me a good thing to have on the contrary:
https://www.nngroup.com/articles/magnifying-glass-icon/

Comment (ArthurHoaro) [inline tpl/default/css/shaarli.css]: * I don't reproduce it on this computer, I'll test it again later. Maybe I was using an outdated version of your PR.
* I was actually suggesting to put the icon in one block at the right of both field, but if you feel like adding a third row is a better idea, go ahead.

> It seems to me that this icon a good thing to have on the contrary

Yes, that what's I meant. Removing the button isn't a good thing regarding accessibility.

**Comment by Lucas-C** [comment]
I fixed the button position for small devices.
What do you think ?

**Comment by nodiscc** [comment]
@Lucas-C your rebased commit causes phpunit tests to fail https://travis-ci.org/shaarli/Shaarli/jobs/241951954#L331

> fixed the button position for small devices.

I just made a quick improvement to the form/button in my browser dev tools, would you be ok to include something similar in your patch? If you like the idea

```html
<!-- add a new div around the button -->
<div id="searchbuttonbox" style="background-color:   #1B926C;margin: 2px 0 3px 0;padding: 3px;border-radius: 2px;">
	<button type="submit" class="search-button" style="color: #fff;"><i class="fa fa-search"></i></button>
</div>
```


sha-search-button-before png
sha-search-button-2

```html
<button type="submit" class="search-button" style="background-color: #1B926C;padding: 4px 8px 6px 8px;border-radius: 2px;color: #fff;"><i class="fa fa-search"></i></button>
```

sha-search-wide-before
sha-search-wide-2

**Comment by Lucas-C** [comment]
The tests failing are due to the changes in `application/LinkFilter.php`, discussed above in this PR.

@ArthurHoaro : what do you think ?
- I change the tests so that they pass, but this commit will remove the "untagged tags" feature
- I include in this PR the additional filter/icon we discussed

About the suggestion to change the look of the search button, I agree it looks better this way !
I'm going to include your modifications.

**Comment by ArthurHoaro** [comment]
I'd prefer that you add the additional filter in this PR to avoid breaking existing feature (being able to look for untagged links). Note that currently, you can use the API to get only untagged links with `searchtags` to `false`.

You also added JS that shouldn't be here with your last rebase.

**Comment by Lucas-C** [comment]
Done

**Comment by Lucas-C** [comment]
I think I fixed the tests.
Are you ok to merge this ?

File: index.php [hunk 9]
```diff
@@ -1651,7 +1665,7 @@ function buildLinkList($PAGE,$LINKSDB, $conf, $pluginManager)
             'searchtags' => $searchtags,
             'searchterm' => $searchterm,
         ];
-        $linksToDisplay = $LINKSDB->filterSearch($request, false, $visibility);
+        $linksToDisplay = $LINKSDB->filterSearch($request, false, $visibility, $_SESSION['untaggedonly']);
```
Comment (ArthurHoaro) [inline index.php]: You need to make sure that the array key exists, otherwise it generates a warning (use `empty()`).

File: application/PageBuilder.php [hunk 10]
```diff
@@ -78,6 +78,7 @@ private function initialize()
         $this->tpl->assign('version', shaarli_version);
         $this->tpl->assign('scripturl', index_url($_SERVER));
         $this->tpl->assign('privateonly', !empty($_SESSION['privateonly'])); // Show only private links?
+        $this->tpl->assign('untaggedonly', $_SESSION['untaggedonly']);
```
Comment (ArthurHoaro) [inline application/PageBuilder.php]: You need to make sure that the array key exists, otherwise it generates a warning (use `empty()`).

File: index.php [hunk 11]
```diff
@@ -1017,6 +1018,19 @@ function renderPage($conf, $pluginManager, $LINKSDB, $history)
         exit;
     }
 
+    // -------- User wants to see only untagged links (toggle)
+    if (isset($_GET['untaggedonly'])) {
+        $_SESSION['untaggedonly'] = !$_SESSION['untaggedonly'];
```
Comment (ArthurHoaro) [inline index.php]: You need to make sure that the array key exists, otherwise it generates a warning (use `empty()`).

Comment (Lucas-C) [inline index.php]: I don't think it's needed there, as I call `isset`

**Comment by Lucas-C** [comment]
Fixes made :)

**Comment by virtualtam** [comment]
Thanks @Lucas-C for the PR ;-)

File: index.php (hunk 11 above)
Comment (ArthurHoaro) [inline index.php]: Actually it was, because you're using `isset` on `$_GET` and accessing `$_SESSION`.

Comment (Lucas-C) [inline index.php]: Oh silly me !
Should I send a PR to fix this ?

Comment (ArthurHoaro) [inline index.php]: Sure!
